### PR TITLE
Add Document Handler

### PR DESCRIPTION
**Pull request by 10eMyrT**
This is linked to the issue #248 (Creation of the document handler). Currently, it just sends a message notifying that a document is received, but it also contains a variable named Link which will contain a URL, once we decide where to host the documents. The file compiles, but since my environment has not been successfully set, I have not sent a document to test the handler. Please let me know if it does not work in any way.

### CODE REVIEW

**Comment by jwflory** [comment]
_[Discussed in 2020-03-14 meeting](https://bluejeans.com/s/O5ssq/)._

---

I did a walk-through with @10eMyrT on feature branching and `git rebase` after the developer meeting. This PR is blocked, for now, on @Tjzabel's Update handler PR. In the meanwhile, @10eMyrT will work on getting the git history cleaned up on this PR.

**Comment by 10eMyrT** [comment]
Apologies on the mess with closing an reopening. I unknowingly closed when I was cleaning the history.

File: cmd/teleirc.go [hunk 1]
```diff
@@ -18,42 +17,43 @@ const (
 
 var (
 	flagPath    = flag.String("conf", ".env", "config file")
-	flagDebug   = flag.Bool("debug", false, "enable debugging")
+	flagDebug   = flag.Bool("debug", false, "disable debugging")
```
Comment (Tjzabel) [inline cmd/teleirc.go]: I think keeping `enable debugging` is better, since this lets users know that any changes to this `Bool` adds debugging capabilities.

File: internal/config.go [hunk 2]
```diff
@@ -18,7 +19,8 @@ type IRCSettings struct {
 	Port                int      `env:"IRC_PORT" envDefault:"6667" validate:"min=0,max=65535"`
 	TLSAllowSelfSigned  bool     `env:"IRC_CERT_ALLOW_SELFSIGNED" envDefault:"true"`
 	TLSAllowCertExpired bool     `env:"IRC_CERT_ALLOW_EXPIRED" envDefault:"true"`
-	Channel             string   `env:"IRC_CHANNEL,required"`
+	Channel             string   `env:"IRC_CHANNEL,required" validate:"notempty"`
+	ChannelKey          string   `env:"IRC_CHANNEL_KEY" envDefault:""`
```
Comment (Tjzabel) [inline internal/config.go]: This seems to be a part of another PR?

File: internal/debug.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+*/
+package internal
+
+import (
+	"log"
+	"os"
+)
+
+var (
+	// logFlags    = log.Ldate | log.Ltime | log.Lshortfile
```
Comment (Tjzabel) [inline internal/debug.go]: This comment seems like it can be removed

File: internal/debug.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func (d Debug) LogInfo(message string) {
+	if d.DebugLevel {
+		info.Println(message)
+	}
+}
+
+func (d Debug) LogError(message error) {
+	errorLog.Println(message)
```
Comment (Tjzabel) [inline internal/debug.go]: Does `Println` take into account the type of output stream to use?

File: internal/handlers/telegram/documentHandler.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
+package telegram
+
+import (
+	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api"
+)
+
+
+type DocumentObject struct {
+	EnableLink bool
```
Comment (Tjzabel) [inline internal/handlers/telegram/documentHandler.go]: WIth the v2.0 iteration of TeleIRC, I would like to move away from enabling any sort of boolean for sending documents.

File: internal/handlers/telegram/documentHandler.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	Link string
+	Object *tgbotapi.Document
```
Comment (Tjzabel) [inline internal/handlers/telegram/documentHandler.go]: Instead of calling this Object, we can call this something a little nicer like `Document`.

File: internal/handlers/telegram/documentHandler.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+/*
+Constructs a new document and is used to decide under which format to formard
```
Comment (Tjzabel) [inline internal/handlers/telegram/documentHandler.go]: This relates to the issue we have up on how to do proper development documentation. GoDoc uses comments to generate dev docs. Start function comments with "functionName does...."

In this case, we can start off with "newDocument constructs a new document and is..."

File: internal/handlers/telegram/documentHandler.go [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+	return DocumentObject{enabled, link, doc}
+}
+
+/*
+documentHandler once a document received, uses information about the document
+to format a message to return to the client.
+*/
+func (doc DocumentObject) documentHandler()  string {
+
+	message := " shared a file (" +
```
Comment (Tjzabel) [inline internal/handlers/telegram/documentHandler.go]: We need to display the Telegram username as well. This doesn't tell us _who_ shared the file.

File: internal/handlers/telegram/documentHandler.go [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+		doc.Object.MimeType + ") on Telegram with title" +
+		"'" + doc.Object.FileName + "'."
+
+	if doc.EnableLink {
```
Comment (Tjzabel) [inline internal/handlers/telegram/documentHandler.go]: As I mentioned earlier, I think we should remove the boolean from this.

File: internal/handlers/telegram/documentHandler.go [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+		message += " View document at " + doc.Link
```
Comment (Tjzabel) [inline internal/handlers/telegram/documentHandler.go]: Since we're only sending over that someone sent a document in Telegram, we don't need to include the link here.

File: cmd/teleirc.go [hunk 11]
```diff
@@ -18,42 +17,43 @@ const (
...
 
+	verbose := internal.Debug { DebugLevel:  *flagDebug }
+
 	if *flagVersion {
-		fmt.Printf("Current TeleIRC version: %s\n", version)
+		verbose.PrintVersion("Current TeleIRC version: " + version)
 		return
 	}
 
 	// TODO: Build out debugging capabilities for more verbose output
```
Comment (Tjzabel) [inline cmd/teleirc.go]: We can remove this comment.

File: cmd/teleirc.go [hunk 12]
```diff
@@ -18,42 +17,43 @@ const (
...
 	flagPath    = flag.String("conf", ".env", "config file")
-	flagDebug   = flag.Bool("debug", false, "enable debugging")
+	flagDebug   = flag.Bool("debug", false, "disable debugging")
 	flagVersion = flag.Bool("version", false, "displays current version of TeleIRC")
 )
 
 func main() {
 	flag.Parse()
 
+	verbose := internal.Debug { DebugLevel:  *flagDebug }
```
Comment (Tjzabel) [inline cmd/teleirc.go]: I wonder if it's better to just call this `log` or `debug` instead of verbose, so it's explicitly clear this is part of the logging.
I can go either way with this, just a thought.

File: internal/debug.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
+/*
+	Debug contains information about debug levels and define
```
Comment (Tjzabel) [inline internal/debug.go]: Not sure why the comments here are indented

File: internal/debug.go [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"log"
+	"os"
+)
+
+var (
+	// logFlags    = log.Ldate | log.Ltime | log.Lshortfile
+	logFlags    = log.Ldate | log.Ltime
+	info = log.New(os.Stdout, "INFO: ", logFlags)
```
Comment (Tjzabel) [inline internal/debug.go]: This should be indented properly with the rest

File: internal/handlers/irc/irc_test.go [hunk 15]
```diff
@@ -36,7 +39,10 @@ func TestNewClientFull(t *testing.T) {
 		BotName:          "test_name",
 		NickServPassword: "test_pass",
 	}
-	client := NewClient(ircSettings)
+	logger := internal.Debug{
+		DebugLevel:  false,
+	}
+	client := NewClient(ircSettings, logger)
```
Comment (Tjzabel) [inline internal/handlers/irc/irc_test.go]: Do these tests pass?

**Comment by Tjzabel** [comment]
_**EDIT**_:  It looks like this PR encompasses 2-3 others. I've added feedback for both the logging PR and this document PR into this one. It seems I was not able to review the logging PR before it was merged in.

File: internal/handlers/telegram/documentHandler.go (hunk 6 above)
Comment (10eMyrT) [reply internal/handlers/telegram/documentHandler.go]: Sure, I like the idea.

File: internal/handlers/irc/irc_test.go (hunk 15 above)
Comment (10eMyrT) [reply internal/handlers/irc/irc_test.go]: Yes, all the tests pass.

File: internal/config.go (hunk 2 above)
Comment (10eMyrT) [reply internal/config.go]: Yes, it is. In this PR, I have not changed anything in this file.

File: internal/debug.go (hunk 4 above)
Comment (10eMyrT) [reply internal/debug.go]: `Println` simply prints the error message to stdout, as it does with `fmt`. Is there an additional need?

Comment (Tjzabel) [reply internal/debug.go]: @10eMyrT we may want this to print out to standard error, not standard out.

Comment (10eMyrT) [reply internal/debug.go]: I see. That's the best way. I will integrate it in the changes before next meeting.

**Comment by jwflory** [comment]
_Discussed in [2020-03-21 meeting](https://bluejeans.com/s/3YKnj/)._

---

@10eMyrT is going to split this up into two smaller PRs from separate feature branches: one for the logging changes and one for the document handler.
